File: biddo_client/lib/core/repositories/review.dart
import 'package:get/get.dart';
import '../models/review.dart';
import 'base.dart';

class ReviewRepository {
  var dio = Get.find<Api>();

  Future<String?> translate(String id, String lang) async {
    try {
      var response = await dio.api.get('/review/translate/review/$id/$lang');
      return response.data['description'];
    } catch (error) {
      return null;
    }
  }

  Future<List<Review>> loadForAccount(
      String accountId, int page, int perPage) async {
    try {
      var response = await dio.api.get('/review/$accountId/$page/$perPage');
      return List<Review>.from(response.data.map((el) => Review.fromJSON(el)));
    } catch (error) {
      return [];
    }
  }

  Future<Review?> saveReview(
    double stars,
    String auctionId, [
    String? description,
    String? id,
  ]) async {
    try {
      var result = await dio.api.post('/review', data: {
        'id': id,
        'stars': stars,
        'description': description,
        'auctionId': auctionId,
      });

      return Review.fromJSON(result.data);
    } catch (error) {
      return null;
    }
  }
}

File: core/repositories/review.ts
import { Review } from '../domain/review'
import { RequestMaker, RequestType } from '../services/request-maker'

class ReviewRepository {
  private basePath = '/review'

  public async translateReview(
    reviewId: string,
    language: string
  ): Promise<{ description: string } | null> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `${this.basePath}/translate/review/${reviewId}/${language}`,
        method: RequestType.GET,
      })) as Record<string, unknown>
      return response as { description: string }
    } catch (error) {
      console.error('Error translating review:', error)
      return null
    }
  }

  public async loadForAccount(accountId: string, page: number, perPage: number): Promise<Review[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `${this.basePath}/${accountId}/${page}/${perPage}`,
        method: RequestType.GET,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Review.fromJSON(el))
    } catch (error) {
      console.error('Error loading reviews for account:', error)
      return []
    }
  }

  public async loadForLoggedAccount(page: number, perPage: number): Promise<Review[]> {
    try {
      const response = (await RequestMaker.makeRequest({
        path: `${this.basePath}/${page}/${perPage}`,
      })) as Record<string, unknown>[]
      return response.map((el: Record<string, unknown>) => Review.fromJSON(el))
    } catch (error) {
      console.error('Error loading reviews for logged account:', error)
      return []
    }
  }

  public async saveReview(params: {
    reviewId?: string
    auctionId: string
    stars: number
    description?: string
  }): Promise<Review | null> {
    const { reviewId, auctionId, stars, description } = params
    try {
      const result = (await RequestMaker.makeRequest({
        path: this.basePath,
        method: RequestType.POST,
        payload: JSON.stringify({
          ...(reviewId ? { id: reviewId } : {}),
          stars: stars,
          description: description,
          auctionId: auctionId,
        }),
        contentType: 'application/json',
      })) as Record<string, unknown>
      return Review.fromJSON(result)
    } catch (error) {
      console.error('Error saving review:', error)
      return null
    }
  }
}

const reviewRepository = new ReviewRepository()
export { reviewRepository as ReviewRepository }
